### PR TITLE
Initial integration with watches-cli tool

### LINKED ISSUE #396: Collect custom JSON data from HTTP REST API and create charts from it
I need to collect data from HTTP REST API providing metrics data in JSON format and create charts from it in the end of pbench run. I am not sure I know how to implement this ATM. Help is appreciated.

What I would like to achieve:
- collect JSON data in periodic fashion during pbench run
- pre-process JSON data if needed
  - i.e. some JSON data may not contain timestamp, in this case a timestamp needs to be attached to this data (the best option is adding a new timestamp field to the root JSON object)
- store this data (somewhere)
- post-process this data and generate individual metrics charts

In the beginning I am fine with implementing something hardcoded (ad-hoc) but in the future I can see how this can be generalized in several useful ways:
- make the REST API to collect from configurable. For example if we wanted to collect metics from Elasticsearch node we should be able to configure which REST endpoints to collect data from.
- make post-processing configurable. This means we should be able to define what values to pull from stored JSON data and how to transform it so that charts can be easily generated from it. Typically, each JSON object will contain some timestamp at the top level and several metrics somewhere deeper.
- similarly make pre-processing configurable

### PR DESCRIPTION
**Pull request by lukas-vlcek**
Initial PR. Closes #396 

Introducing integration with `watches` cli tool by means of implementing three pbench collection tools: `watches-cluster-health`, `watches-cluster-stats` and `watches-nodes-stats`. There can be more tools implemented to collect more valuable data from ES but this should be good start.

Individual post-process scripts should work correctly, I did quite a solid testing locally. If there can be implemented any automated tests for this, please let me know.

Still, I need help with reviewing changes to `kvm-spinlock` script. I am not sure this will work correctly.

Speaking about post-process scripts they can be resource intensive when generated HTML is opened in web browser. For example I polled stats from 3 node cluster and the nodes-stats HTML page contains almost 250 charts. If I understand correctly, each of those charts do async AJAX call to load the csv data.

### CODE REVIEW

**Comment by lukas-vlcek** [comment]
@ndokos @portante @richm PTAL

**Comment by lukas-vlcek** [comment]
One of the things I need to improve is enabling conditional use of `watches -s` option in all `*-datalog` scripts. In some execution contexts (like when ES is running in Openshift) the `watches` tool can not route HTTP(S) requests to individual ES nodes but has to send all requests through explicit proxy or router. Thus `-s` option does not apply in this case and **must not** be used.

**Comment by richm** [comment]
looks good
There is a lot of code duplication among the postprocess scripts.

**Comment by lukas-vlcek** [comment]
@portante @ndokos I would welcome if you can review this PR. I am sure there is a room for improvements at least in my modifications to `agent/tool-scripts/kvm-spinlock` script.

Can you please also point me to some resources how to implemented tests? I haven't investigated pbench tests much so far.

@rflorenc Would you be able to give this a try? Feel free to ping me anytime.

**Comment by portante** [comment]
@lukas-vlcek, on my backlog, we have indexing work we are focusing on first, though.

File: agent/tool-scripts/kvm-spinlock [hunk 1]
```diff
@@ -35,10 +35,22 @@ case "$script_name" in
...
+	watches-cluster-health|watches-cluster-stats|watches-nodes-stats)
+	url="http://localhost:9200"
+	sniff=false # do not use ES sniffing by default
+	buffer=true # use python buffering by default
+	# If Basic auth is used then provide username and password
+	username=""
+	password=""
+	# http header parameters are provided they are used for TCL security
+	http_header_username=""
+    http_header_password=""
```
Comment (portante) [inline agent/tool-scripts/kvm-spinlock]: Spacing issue here of some sort...

File: agent/tool-scripts/kvm-spinlock [hunk 2]
```diff
@@ -235,6 +257,9 @@ case "$tool" in
 	rabbit)
 	tool_cmd="$script_path/datalog/$tool-datalog $interval $username $password $tool_stdout_file"
 	;;
+	watches-cluster-health|watches-cluster-stats|watches-nodes-stats)
+	tool_cmd="$script_path/datalog/$tool-datalog $interval $tool_stdout_file $url $sniff $buffer $username $password $http_header_username $http_header_password $http_header_token"
```
Comment (portante) [inline agent/tool-scripts/kvm-spinlock]: The fact that we have one line here supports the notion that we should have one script passing the appropriate parameter to get the desired behavior.

File: agent/tool-scripts/postprocess/watches-cluster-health-postprocess [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+close(TOOL_TXT);
+
+# define the graph types
+# if you want something other than lineChart, put it here
+my %graph_type;
+
+# threshold for displaying a series in a graph
+my %graph_threshold;
+
+gen_data(\%stats, \%graph_type, \%graph_threshold, $dir);
```
Comment (portante) [inline agent/tool-scripts/postprocess/watches-cluster-health-postprocess]: Missing newline ...

File: agent/tool-scripts/postprocess/watches-cluster-health-postprocess [hunk 4]
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/perl
+
+# Author: Lukas Vlcek (lvlcek@redhat.com)
+#
+# usage: <tool>-postprocess <dir>  dir = directory where tool txt file can be found
+#
+# The purpose of this script is to
+# -output the chart data in JSON format (watches.js)
```
Comment (portante) [inline agent/tool-scripts/postprocess/watches-cluster-health-postprocess]: It seems that the purpose of a post-processing tool is to convert the generated output of the tool into the appropriate machine and human-readable forms.  However, it seems that the `GenData` model actually emits transformed out, and this module just reads it in.  Perhaps we can say just that?

Comment (lukas-vlcek) [inline agent/tool-scripts/postprocess/watches-cluster-health-postprocess]: I took this comment from other postprocessing script (I think it was `pidstat-postprocess`, but similar _formula_ can be found in may other scripts). May be we should open more general ticket for this?

File: agent/tool-scripts/postprocess/watches-cluster-health-postprocess (hunk 3 above)
Comment (lukas-vlcek) [inline agent/tool-scripts/postprocess/watches-cluster-health-postprocess]: I added new line to all three watches postprocessing scripts.

File: agent/tool-scripts/postprocess/unittests [hunk 5]
```diff
@@ -191,14 +191,14 @@ for i in $(ls -1 samples/) ; do
...
             sts=$?
             ;;
-	process-iteration-samples-*)
-	    testprocessiterationsamples $i
-	    sts=$?
-	    ;;
-	generate-benchmark-summary-*)
-	    testgeneratebenchmarksummary $i
-	    sts=$?
-	    ;;
```
Comment (lukas-vlcek) [inline agent/tool-scripts/postprocess/unittests]: Seems there is spaces-vs-tabs mishmash

**Comment by lukas-vlcek** [comment]
@portante Is it possible to run pbench tests with Python 2.7? It seems that currently travis tests run using Python 3.5 and that is why installation of watches fails. I think this is because Python 3.5 is not backward compatible with 2.7.

//cc @richm do you think we can assume watches will have Python 3.5 available? If yes I can update watches to newer version of Python.

**Comment by ndokos** [comment]
@lukas-vlcek: we need python 3.5 to run the indexing tests, but it is installed as python3. Plain python should still be python 2.7. The errors might be coming from pip however: IIRC, a python3 version is installed specifically.

Can you change the order of operations in the .travis.yml file and try again? Put the `sudo pip install watches` before all the python3 stuff. I did a small test and it seems to work but YMMV.

**Comment by lukas-vlcek** [comment]
@ndokos thanks, I will give it a try, if it does not help may be we could use sixer? See https://github.com/ViaQ/watches-cli/issues/34

**Comment by ndokos** [comment]
The travis-ci environment is Ubuntu 12.04 :-(  and it's a bear to debug when something goes wrong. In the past I've set up a local VM to test things out, but in general I try to avoid messing with it.

**Comment by lukas-vlcek** [comment]
Seems we might be running into similar problem: http://stackoverflow.com/questions/32386469/logging-module-not-working-with-python3

**Comment by lukas-vlcek** [comment]
@ndokos we are adding Python 3.5 support to watches, see https://github.com/ViaQ/watches-cli/issues/34 I think it should be doable.

File: agent/tool-scripts/kvm-spinlock (hunk 2 above)
Comment (lukas-vlcek) [inline agent/tool-scripts/kvm-spinlock]: Done

**Comment by lukas-vlcek** [comment]
@portante @ndokos @richm PTAL again

I think I addressed all main issues and comments now:
- `watches` is now [compatible with Python 3.5](https://github.com/ViaQ/watches-cli/releases/tag/v1.0.2)
- reduced three `datalog` tools to common single script `watches-cli-datalog`
- added postprocess tests (samples and gold) which increased the PR size dramatically (in terms of number of files, even GitHub web UI is now struggling in making navigation useful&easy)

I think there is a lot of room for improvements, especially in displaying detailed metric charts in more useful/readable way. But may be we shall address these improvements in new PRs and not making this PR bigger?

**Comment by lukas-vlcek** [comment]
Oh... and we shall probably add some documentation... However, if this means we need to modify `gh-pages` branch then I think I have to open separate PR for docs anyway. Or am I missing anything?

**Comment by lukas-vlcek** [comment]
@ndokos @portante ping, I would like to help merge this is there are no other review blockers.

**Comment by mffiedler** [comment]
@ndokos @portante Any blockers preventing merge?    /cc: @lukas-vlcek

**Comment by lukas-vlcek** [comment]
Just notice: I expect that after merge there might be needed additional polishing to make the integration with "ES inside the OpenShift use case" smooth. Or is it possible to provide integration tests for this scenario as well?
_This ^^ is heads up in case you would like to give this code a try soon._

**Comment by portante** [comment]
I am not in favor of merging this PR for a few reasons:

1. Adding a specific tool like watches, where that tool could be integrated using other supported mechanisms, like the prometheus-metrics tool, or the user-tool functionality, seems to add another support burden on this code base
   * What was initially described as a general tool to handle arbitrary JSON data in #396 seems to have been turned into a tool specifically for handling `watches` data output
2. The PR itself as it currently stands adds 300+ files to the code base, most of them being gold files
   * The samples should only be sufficient to exercise the various code paths written in the tooling
3. There are an existing tools for collecting metrics from elasticsearch, for example [pcp's elasticsearch module](https://github.com/Aconex/elasticsearch-pcp) and [collectd's elasticsearch plugin](https://github.com/signalfx/integrations/tree/master/collectd-elasticsearch)
   * We have support for collecting metrics from pcp
   * We'd love a PR that adds support for collecting metrics from collectd

**Comment by lukas-vlcek** [comment]
@portante 
My reply to 1) can be found here: https://github.com/distributed-system-analysis/pbench/issues/396#issuecomment-307585887

As for 2) this can be easily fixed and we can add recommendations for pbench developers about how to contribute tests in their PRs for the future.

As for 3) to me the pcp ES module does not seem to be maintained. As for collectd, please see again my response here: https://github.com/distributed-system-analysis/pbench/issues/396#issuecomment-307585887

**Comment by portante** [comment]
Would rather see us use the new prometheus metrics end points for Elasticsearch so that we can take advantage of the existing prometheus-metrics collection tool already in pbench.

**Pull request closed without merging**